app/v4/src/index.ts

```diff
@@ -1,9 +1,65 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the MIT License.
 
-import { registerHook } from '@azure/functions-core';
+import { app } from '@azure/functions';
 
-registerHook('postInvocation', async () => {
+app.hook.appStart((context) => {
+    context.hookData.testData = 'appStartHookData123';
+    console.log('appStart hook executed.');
+});
+
+app.hook.appTerminate((context) => {
+    console.log(`appTerminate hook executed with hook data ${context.hookData.testData}`);
+});
+
+app.hook.preInvocation((context) => {
+    if (context.invocationContext.functionName === 'httpTriggerForHooks') {
+        context.invocationContext.log(
+            `preInvocation hook executed with inputs ${JSON.stringify(context.inputs.map((v) => v.constructor.name))}`
+        );
+
+        context.inputs.push('extraTestInput12345');
+
+        const oldHandler: any = context.functionHandler;
+        context.functionHandler = (...args) => {
+            context.invocationContext.log('extra log from updated functionHandler');
+            return oldHandler(...args);
+        };
+
+        context.hookData.testData = 'preInvocationHookData123';
+
+        // Validate readonly properties
+        try {
+            // @ts-expect-error by-design
+            context.hookData = {};
+        } catch (err) {
+            context.invocationContext.log(`Ignored error: ${err.message}`);
+        }
+
+        try {
+            // @ts-expect-error by-design
+            context.invocationContext = {};
+        } catch (err) {
+            context.invocationContext.log(`Ignored error: ${err.message}`);
+        }
+    }
+});
+
+const hookToDispose = app.hook.preInvocation((context) => {
+    context.invocationContext.log('This should never run.');
```

Comment:
Should we add something to this log to make it easier to spot? I assume we should never see it logged if things go right, so adding some extra characters (exclamation points? all caps? stars?) might make it pop more in the console when it does appear for some reason.

Reply:
I added this log to explain what the hook is doing, but I feel like it's ineffective to validate the non-existence of a log. There's so many bugs that could make the log go missing and thus render the validation meaningless. Instead, the main way I see the dispose being validated is through the following:
```
await waitForOutput(`Executing 1 "preInvocation" hooks`);
```
If it didn't work, we would see `2`.